README.md
# b8 - bayesian spam filter in PHP

This is a mirror of the source code of b8 bayesian spam filter, from http://nasauber.de/opensource/b8/ (down at the moment).

The only difference with the mainline release is that this version supports SQLite as a storage backend as well.

## What is b8?

b8 is a statistical ("Bayesian" [1]) spam filter implemented in PHP 5. It is intended to keep your weblog or guestbook spam-free. The filter can be used anywhere in your PHP code and tells you whether a text is spam or not, using statistical text analysis. What it does is: you give b8 a text and it returns a value between 0 and 1, saying it's ham when it's near 0 and saying it's spam when it's near 1. See How does it work? for details about this.

Principally, It's a program like Bogofilter or SpamBayes, but it is not intended to classify emails. Therefore, the way b8 works is slightly different from email spam filters. See What's different? if you're interested in the details.

To be able to distinguish spam and ham (non-spam), b8 first has to learn some spam and some ham texts. If it makes mistakes when classifying unknown texts or the result is not distinct enough, b8 can be told what the text actually is, getting better with each learned text.

The whole documentation can be found in the readme.

Big thanks go to Gary Robinson, as his articles A Statistical Approach to the Spam Problem and Spam Detection describe the foundation for the math and algorithms used in b8.

[1] I'm not a mathematician, but as far as I can grasp it, the math used in b8 has not much to do with Bayes' theorem itself. So I call it a statistical spam filter, not a Bayesian spam filter. 


doc/ChangeLog
2012-11-30 Tobias Leupold <tobias.leupold@web.de>

	* Release: Version 0.6
	
	* Lots of changes (new major release!):
		- Finally did an actually really complete abstraction of the storage backends, so that it was possible to change MySQL's database layout to multiple columns.
		- Kicked out the never-used lastseen parameter.
		- Renamed the internal variables to "b8*...", combined "bayes*text.ham" and "bayes*texts.spam" to "b8*texts".
		- Thus, the database layout has been changed. Included update scripts for both MySQL and DBA.
		- Removed all validate() functions in favor of throwing exceptions when something's wrong.
		- Made the lexer more flexible. Added functions for all split work, that, except for the raw split, can be turned on and off via a new config array.
		- The lexer now supports getting BBCode.
		- Added an additional check to the lexer to be sure no token will collide with an internal variable
		- Added multibyte support to the degenerator so it is now able to handle non-latin-1 texts in the same way as it handles latin-1-texts.
		- b8's constructor now takes four config arrays, the third is the lexer config, the fourth is the degenerator config.
		
	* Bugfixes:
		- Removed the ucfirst function from the degenerator and replaced it with a custom one. It did not what I always thought it would do (first letter upper case, rest lower case), but does only converted the first letter to upper case.
		- Fixed the MySQL backend so it's now able to handle a get() request for an empty array or an array containing just one token.
		- Fixed the MySQL backend when doing a query with no returned result.
		- Fixed the lexer to never output an empty array of tokens, but a placeholder token if no token has been found.
		
2011-04-19 Tobias Leupold <tobias.leupold@web.de>

	* Release: Version 0.5.2
	
	* storage/storage_mysql.php: The destructor of the MySQL storage backend always disconnected the MySQL connection, even if an existing link-resource was passed to b8. Now, the MySQL connection is only disconnected if b8 created the resource itself. Thanks to Gerhard Waldemair for the bug report!
	
2010-12-30 Tobias Leupold <tobias.leupold@web.de>

	* Release: Version 0.5.1
	
	* Bigger changes:
		- Fixed some issues with the scope of variables leading to problems when multiple instances of b8 are created. Thanks to Mike Creuzer for the bug report :-)
		- Centralized the loading of class definition files in the b8 constructor and created a function to handle the inclusion.
		
	* b8.php: Return a lexer error code instead of a rating if the lexer failed. The lexer never returned FALSE but b8 checked only for this value to validate the lexer didn't fail. Thanks to Matt Friedman for the bug report :-)
	
	* lexer/lexer_default.php: A bit of code cleanup: less useless nesting.
	
	* doc/readme.*: Updated the documentation, added a FAQ.
	
2010-06-27 Tobias Leupold <tobias.leupold@web.de>

	* Release: Version 0.5-r1
	
	* doc/readme.*: Updated the documentation; forgot the newly introduced b8::HAM and b8::SPAM variables. Added some additional information about the storage model.
	
2010-06-02 Tobias Leupold <tobias.leupold@web.de>

	* Release: Version 0.5
	
	* 100.000 Changes (new major release!), at a glance:
		- No PHP 4 compatibility anymore. Much cleaner code base with less hacks.
		- Completely reworked storage model. The SQL performance increased dramatically, the Berkeley DB performance remains as fast as it always has been.
		- Better lexer which can also handle non-latin1 texts in a nice way, so that e. g. Cyrillic or Chinese texts can be classified more performant.
		- No config files anymore, multiple instances of b8 can be now created in the same script with different configuration, databases and no problems.
		- No spooky administration interface anymore that needs an SQL database, even if Berkeley DB is used (anybody who actually used this?! I never did ;-).
		- No "install" scripts and routines and a less end-user compatible documentation. Anybody integrating b8 in his homepage won't be an end-user, will he?
		
2009-02-03 Oliver Lillie (aka buggedcom)

	* Revision: 221 (the original PHP 5 port)
	
	* Rewrote Tobias' original class for optimisation and PHP 5 functionality.
	
	* Improved database mysql query useage by over ~820%
	
	* Class is faster, ~20%.
	
	* Slight increase in memory usage, but it's small and given the advantages of the speed increase and query reduction it's worth it.
	
	* Removed install code from mysql class and added a sql file. Anyone who wants to use this is generally going to be more advanced anyway and see the sql to install.
	
2009-02-03 Tobias Leupold <tobias.leupold@web.de>

	* Release: Version 0.4.4 -- changed the license type from GPL to LGPL
	
2008-06-27 Tobias Leupold <tobias.leupold@web.de>

	* Release: Version 0.4.3 -- no bugs found ... so let's make a release with only small changes ;-)
	
	* b8.php: Removed debugging messages that were commented out anyway
	
	* storage/storage_mysql.php: Made it possible to pass both a MySQL-link resource and a table name to b8. This makes b8 useable in the Redaxo CMS (and probably others)
	
	* doc/readme.htm: Updated documentation accordingly
	
2008-02-17 Tobias Leupold <tobias.leupold@web.de>

	* Release: Version 0.4.2
	
	* interface/backup.php: the bayes*dbversion tag is now written to a database emptied by drop(), so that it will be useable without an error message even if no backup is recovered afterwards.
	
	* doc/readme.htm: added a security note to the configuration section (htaccess should be used to avoid everybody to be able to see the configuration)
	
2007-09-17 Tobias Leupold <tobias.leupold@web.de>

	* Release: Version 0.4.1
	
	* storage/storage_mysql.php: fixed b8 crashing when getting passed a persistent MySQL resource link. Thanks to Paul Chapman for the bug report :-)
	
2007-06-08 Tobias Leupold <tobias.leupold@web.de>

	* Release: Version 0.4
	

b8/storage/storage_mysql.php
<?php

#   Copyright (C) 2006-2012 Tobias Leupold <tobias.leupold@web.de>
#
#   This file is part of the b8 package
#
#   This program is free software; you can redistribute it and/or modify it
#   under the terms of the GNU Lesser General Public License as published by
#   the Free Software Foundation in version 2.1 of the License.
#
#   This program is distributed in the hope that it will be useful, but
#   WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
#   or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
#   License for more details.
#
#   You should have received a copy of the GNU Lesser General Public License
#   along with this program; if not, write to the Free Software Foundation,
#   Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

/**
 * The MySQL backend for communicating with the database.
 * Copyright (C) 2009 Oliver Lillie
 * Copyright (C) 2010-2012 Tobias Leupold <tobias.leupold@web.de>
 *
 * @license LGPL 2.1
 * @access public
 * @package b8
 * @author Oliver Lillie (original PHP 5 port and optimizations)
 * @author Tobias Leupold
 */

class b8_storage_mysql extends b8_storage_base
{

	public $config = array(
		'database'   => 'b8_wordlist',
		'table_name' => 'b8_wordlist',
		'host'       => 'localhost',
		'user'       => FALSE,
		'pass'       => FALSE,
		'connection' => NULL
	);
	
	private $_connection = NULL;
	
	private $_deletes = array();
	private $_puts    = array();
	private $_updates = array();
	
	/**
	 * Constructs the backend.
	 *
	 * @access public
	 * @param string $config
	 */
	
	function __construct($config, &$degenerator)
	{
	
		# Pass the degenerator instance to this class
		$this->degenerator = $degenerator;
		
		# Validate the config items
		
		foreach($config as $name => $value) {
		
			switch($name) {
			
				case 'table_name':
				case 'host':
				case 'user':
				case 'pass':
				case 'database':
					$this->config[$name] = (string) $value;
					break;
					
				case 'connection':
					$this->config['connection'] = $value;
					break;
					
				default:
					throw new Exception("b8_storage_mysql: Unknown configuration key: \"$name\"");
					
			}
			
		}
		
		# Connect to the database
		
		if($this->config['connection'] !== NULL) {
		
			# A resource has been passed, so check if it's okay.
			
			if(is_resource($this->config['connection']) === TRUE) {
				
				$resource_type = get_resource_type($this->config['connection']);
				
				if($resource_type != 'mysql link' and $resource_type != 'mysql link persistent')
					throw new Exception('b8_storage_mysql: The resource passed via the "connection" paramter is no MySQL resource.');
			
			}
			else
				throw new Exception('b8_storage_mysql: The resource passed via the "connection" paramter is no resource.');
			
			# If we reach here, we can use the passed resource.
			$this->_connection = $this->config['connection'];
			
		}

		else {
		
			# We have to connect.
			
			$this->_connection = mysql_connect($this->config['host'], $this->config['user'], $this->config['pass']);
			
			if($this->_connection === FALSE or mysql_select_db($this->config['database'], $this->_connection) === FALSE)
				throw new Exception('b8_storage_mysql: ' . mysql_error());
			
		}
		
		# Check to see if the wordlist table exists
		if(mysql_query('DESCRIBE `' . $this->config['table_name'] . '`', $this->_connection) === FALSE)
			throw new Exception('b8_storage_mysql: ' . mysql_error());
		
		# Let's see if this is a b8 database and the version is okay
		$this->check_database();
		
	}
	
	/**
	 * Closes the database connection.
	 *
	 * @access public
	 * @return void
	 */
	
	function __destruct()
	{
	
		# Commit any changes before closing
		$this->_commit();
		
		# Just close the connection if no link-resource was passed and b8 created it's own connection
		if($this->config['connection'] === NULL)
			mysql_close($this->_connection);
		
	}
	
	/**
	 * Does the actual interaction with the database when fetching data.
	 *
	 * @access protected
	 * @param array $tokens
	 * @return mixed Returns an array of the returned data in the format array(token => data) or an empty array if there was no data.
	 */
	
	protected function _get_query($tokens)
	{
	
		# Construct the query ...
		
		if(count($tokens) > 1) {
		
			# We have more than 1 token
			
			$where = array();
			
			foreach ($tokens as $token) {
				$token = mysql_real_escape_string($token, $this->_connection);
				array_push($where, $token);
			}
			
			$where = "token IN ('" . implode("', '", $where) . "')";
		}
		
		elseif(count($tokens) == 1) {
			# We have exactly one token
			$token = mysql_real_escape_string($tokens[0], $this->_connection);
			$where = "token = '" . $token . "'";
		}
		
		elseif(count($tokens) == 0) {
			# We have no tokens
			# This can happen when we do a degenerates lookup and we don't have any degenerates.
			return array();
		}
		
		# ... and fetch the data
		
		$result = mysql_query('
			SELECT token, count_ham, count_spam
			FROM `' . $this->config['table_name'] . '`
			WHERE ' . $where . ';
		', $this->_connection);
		
		# Check if anything matched the query
		if($result === FALSE)
			return array();
		
		$data = array();
		
		while($row = mysql_fetch_assoc($result)) {
			$data[$row['token']] = array(
				'count_ham'  => $row['count_ham'],
				'count_spam' => $row['count_spam']
			);
		}
		
		mysql_free_result($result);
		
		return $data;
		
	}
	
	/**
	 * Store a token to the database.
	 *
	 * @access protected
	 * @param string $token
	 * @param string $count
	 * @return void
	 */
	
	protected function _put($token, $count)
	{
		array_push($this->_puts,
			"('"
			. mysql_real_escape_string($token, $this->_connection)
			. "', '"
			. mysql_real_escape_string($count['count_ham'], $this->_connection)
			. "', '"
			. mysql_real_escape_string($count['count_spam'], $this->_connection)
			. "')"
		);
	}
	
	/**
	 * Update an existing token.
	 *
	 * @access protected
	 * @param string $token
	 * @param string $count
	 * @return void
	 */
	
	protected function _update($token, $count)
	{
		array_push($this->_updates,
			"('"
			. mysql_real_escape_string($token, $this->_connection)
			. "', '"
			. mysql_real_escape_string($count['count_ham'], $this->_connection)
			. "', '"
			. mysql_real_escape_string($count['count_spam'], $this->_connection)
			. "')"
		);
	}
	
	/**
	 * Remove a token from the database.
	 *
	 * @access protected
	 * @param string $token
	 * @return void
	 */
	
	protected function _del($token)
	{
		array_push($this->_deletes, mysql_real_escape_string($token, $this->_connection));
	}
	
	/**
	 * Commits any modification queries.
	 *
	 * @access protected
	 * @return void
	 */
	
	protected function _commit()
	{
	
		if(count($this->_deletes) > 0) {
		
			$result = mysql_query("
				DELETE FROM `{$this->config['table_name']}`
				WHERE token IN ('" . implode("', '", $this->_deletes) . "');
			", $this->_connection);
			
			if(is_resource($result) === TRUE)
				mysql_free_result($result);
			
			$this->_deletes = array();
			
		}
		
		if(count($this->_puts) > 0) {
		
			$result = mysql_query("
				INSERT INTO `{$this->config['table_name']}`(token, count_ham, count_spam)
				VALUES " . implode(', ', $this->_puts) . ';
			', $this->_connection);
			
			if(is_resource($result) === TRUE)
				mysql_free_result($result);
				
			$this->_puts = array();
			
		}
		
		if(count($this->_updates) > 0) {
		
			$result = mysql_query("
				INSERT INTO `{$this->config['table_name']}`(token, count_ham, count_spam)
				VALUES " . implode(', ', $this->_updates) . "
				ON DUPLICATE KEY UPDATE
					`{$this->config['table_name']}`.count_ham = VALUES(count_ham),
					`{$this->config['table_name']}`.count_spam = VALUES(count_spam);
			", $this->_connection);
			
			if(is_resource($result) === TRUE)
				mysql_free_result($result);
				
			$this->_updates = array();
			
		}
		
	}
	
}

?>

b8/b8.php
<?php

#   Copyright (C) 2006-2012 Tobias Leupold <tobias.leupold@web.de>
#
#   b8 - A statistical ("Bayesian") spam filter written in PHP 5
#
#   This program is free software; you can redistribute it and/or modify it
#   under the terms of the GNU Lesser General Public License as published by
#   the Free Software Foundation in version 2.1 of the License.
#
#   This program is distributed in the hope that it will be useful, but
#   WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
#   or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
#   License for more details.
#
#   You should have received a copy of the GNU Lesser General Public License
#   along with this program; if not, write to the Free Software Foundation,
#   Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

/**
 * Copyright (C) 2006-2013 Tobias Leupold <tobias.leupold@web.de>
 *
 * @license LGPL 2.1
 * @access public
 * @package b8
 * @author Tobias Leupold
 * @author Oliver Lillie (original PHP 5 port)
 */

class b8
{

	const DBVERSION = 3;

	public $config = array(
		'lexer'        => 'default',
		'degenerator'  => 'default',
		'storage'      => 'dba',
		'use_relevant' => 15,
		'min_dev'      => 0.2,
		'rob_s'        => 0.3,
		'rob_x'        => 0.5
	);
	
	public $storage     = NULL;
	public $lexer       = NULL;
	public $degenerator = NULL;
	
	private $_token_data = NULL;
	
	const SPAM    = 'spam';
	const HAM     = 'ham';
	const LEARN   = 'learn';
	const UNLEARN = 'unlearn';
	
	const CLASSIFYER_TEXT_MISSING = 'CLASSIFYER_TEXT_MISSING';
	
	const TRAINER_TEXT_MISSING     = 'TRAINER_TEXT_MISSING';
	const TRAINER_CATEGORY_MISSING = 'TRAINER_CATEGORY_MISSING';
	const TRAINER_CATEGORY_FAIL    = 'TRAINER_CATEGORY_FAIL';

	/**
	 * Constructs b8
	 *
	 * @access public
	 * @return void
	 */
	
	function __construct(
		$config             = array(),
		$config_storage     = array(),
		$config_lexer       = array(),
		$config_degenerator = array()
	)
	{
	
		# Validate config data
		
		foreach ($config as $name => $value) {
		
			switch($name) {
			
				case 'min_dev':
				case 'rob_s':
				case 'rob_x':
					$this->config[$name] = (float) $value;
					break;
					
				case 'use_relevant':
					$this->config[$name] = (int) $value;
					break;
					
				case 'lexer':
				case 'degenerator':
				case 'storage':
					$this->config[$name] = (string) $value;
					break;
				
				default:
					throw new Exception("b8: Unknown configuration key: \"$name\"");
			}
			
		}
		
		# Setup the degenerator class
		
		$class = $this->_load_class('degenerator', $this->config['degenerator']);
		
		if($class === FALSE)
			throw new Exception("b8: Could not load class definition file for degenerator \"{$this->config['degenerator']}\"");
		
		$this->degenerator = new $class($config_degenerator);
		
		# Setup the lexer class
		
		$class = $this->_load_class('lexer', $this->config['lexer']);
		
		if($class === FALSE)
			throw new Exception("b8: Could not load class definition file for lexer \"{$this->config['lexer']}\"");
		
		$this->lexer = new $class($config_lexer);
		
		# Setup the storage backend
		
		$class = $this->_load_class('storage', 'base');
		
		if($class === FALSE)
			throw new Exception("b8: Could not load class definition file for the storage base class");
		
		$class = $this->_load_class('storage', $this->config['storage']);
		
		if($class === FALSE)
			throw new Exception("b8: Could not load class definition file for storage backend \"{$this->config['storage']}\"");
		
		$this->storage = new $class($config_storage, $this->degenerator);
		
	}
	
	/**
	 * Load a class file if a class has not been defined yet.
	 *
	 * @access private
	 * @return boolean Returns TRUE if everything is okay, otherwise FALSE.
	 */
	
	private function _load_class($class_type, $class_name)
	{
	
		$complete_class_name = "b8_{$class_type}_{$class_name}";
		$class_file = dirname(__FILE__) . DIRECTORY_SEPARATOR . $class_type . DIRECTORY_SEPARATOR . "{$class_type}_{$class_name}.php";
	
		if(class_exists($complete_class_name, FALSE) === FALSE) {
		
			# Check if the requested file actually exists
			if(is_file($class_file) !== TRUE or is_readable($class_file) !== TRUE)
				return FALSE;
			
			# Include it
			$included = require_once($class_file);
			
			if($included === FALSE or class_exists($complete_class_name, FALSE) === FALSE)
				return FALSE;
				
		}
		
		return $complete_class_name;
		
	}
	
	/**
	 * Classifies a text
	 *
	 * @access public
	 * @param string $text
	 * @return mixed float The rating between 0 (ham) and 1 (spam) or an error code
	 */
	
	public function classify($text = NULL)
	{
	
		# Let's first see if the user called the function correctly
		if($text === NULL)
			return self::CLASSIFYER_TEXT_MISSING;
	
		# Get the internal database variables, containing the number of ham and
		# spam texts so the spam probability can be calculated in relation to them
		$internals = $this->storage->get_internals();
		
		# Calculate the spamminess of all tokens
		
		# Get all tokens we want to rate
		$tokens = $this->lexer->get_tokens($text);
		
		# Check if the lexer failed
		# (if so, $tokens will be a lexer error code, if not, $tokens will be an array)
		if(!is_array($tokens))
			return $tokens;
			
		# Fetch all available data for the token set from the database
		$this->_token_data = $this->storage->get(array_keys($tokens));
		
		# Calculate the spamminess and importance for each token (or a degenerated form of it)
		
		$word_count = array();
		$rating     = array();
		$importance = array();
		
		foreach($tokens as $word => $count) {
		
			$word_count[$word] = $count;
			
			# Although we only call this function only here ... let's do the
			# calculation stuff in a function to make this a bit less confusing ;-)
			$rating[$word] = $this->_get_probability($word, $internals['texts_ham'], $internals['texts_spam']);
			
			$importance[$word] = abs(0.5 - $rating[$word]);
			
		}
		
		# Order by importance
		arsort($importance);
		reset($importance);
		
		# Get the most interesting tokens (use all if we have less than the given number)
		
		$relevant = array();
		
		for($i = 0; $i < $this->config['use_relevant']; $i++) {
		
			if($tmp = each($importance)) {
				
				# Important tokens remain
				
				# If the token's rating is relevant enough, use it
				
				if(abs(0.5 - $rating[$tmp['key']]) > $this->config['min_dev']) {
				
					# Tokens that appear more than once also count more than once
					
					for($x = 0, $l = $word_count[$tmp['key']]; $x < $l; $x++)
						array_push($relevant, $rating[$tmp['key']]);
						
				}
				
			}
			
			else {
				# We have less words as we want to use, so we
				# already use what we have and can break here
				break;
			}
			
		}
		
		# Calculate the spamminess of the text (thanks to Mr. Robinson ;-)
		# We set both hamminess and spamminess to 1 for the first multiplying
		$hamminess  = 1;
		$spamminess = 1;
		
		# Consider all relevant ratings
		foreach($relevant as $value) {
			$hamminess  *= (1.0 - $value);
			$spamminess *= $value;
		}
		
		# If no token was good for calculation, we really don't know how
		# to rate this text, so can return 0.5 without further calculations.
		if($hamminess == 1 and $spamminess == 1)
			return 0.5;
		
		# Calculate the combined rating

		# Get the number of relevant ratings
		$n = count($relevant);

		# The actual hamminess and spamminess
		$hamminess  = 1 - pow($hamminess,  (1 / $n));
		$spamminess = 1 - pow($spamminess, (1 / $n));
		
		# Calculate the combined indicator
		$probability = ($hamminess - $spamminess) / ($hamminess + $spamminess);
		
		# We want a value between 0 and 1, not between -1 and +1, so ...
		$probability = (1 + $probability) / 2;
		
		# Alea iacta est
		return $probability;
		
	}
	
	/**
	 * Calculate the spamminess of a single token also considering "degenerated" versions
	 *
	 * @access private
	 * @param string $word
	 * @param string $texts_ham
	 * @param string $texts_spam
	 * @return void
	 */
	
	private function _get_probability($word, $texts_ham, $texts_spam)
	{
	
		# Let's see what we have!
		
		if(isset($this->_token_data['tokens'][$word]) === TRUE) {
			# The token is in the database, so we can use it's data as-is
			# and calculate the spamminess of this token directly
			return $this->_calc_probability($this->_token_data['tokens'][$word], $texts_ham, $texts_spam);
		}
		
		# The token was not found, so do we at least have similar words?
		
		if(isset($this->_token_data['degenerates'][$word]) === TRUE) {
		
			# We found similar words, so calculate the spamminess for each one
			# and choose the most important one for the further calculation
			
			# The default rating is 0.5 simply saying nothing
			$rating = 0.5;
			
			foreach($this->_token_data['degenerates'][$word] as $degenerate => $count) {
			
				# Calculate the rating of the current degenerated token
				$rating_tmp = $this->_calc_probability($count, $texts_ham, $texts_spam);
				
				# Is it more important than the rating of another degenerated version?
				if(abs(0.5 - $rating_tmp) > abs(0.5 - $rating))
					$rating = $rating_tmp;
				
			}
			
			return $rating;
			
		}
		
		else {
			# The token is really unknown, so choose the default rating
			# for completely unknown tokens. This strips down to the
			# robX parameter so we can cheap out the freaky math ;-)
			return $this->config['rob_x'];
		}
		
	}
	
	/**
	 * Do the actual spamminess calculation of a single token
	 *
	 * @access private
	 * @param array $data
	 * @param string $texts_ham
	 * @param string $texts_spam
	 * @return void
	 */
	
	private function _calc_probability($data, $texts_ham, $texts_spam)
	{
	
		# Calculate the basic probability as proposed by Mr. Graham
		
		# But: consider the number of ham and spam texts saved instead of the
		# number of entries where the token appeared to calculate a relative
		# spamminess because we count tokens appearing multiple times not just
		# once but as often as they appear in the learned texts.
		
		$rel_ham = $data['count_ham'];
		$rel_spam = $data['count_spam'];
		
		if($texts_ham > 0)
			$rel_ham = $data['count_ham'] / $texts_ham;
			
		if($texts_spam > 0)
			$rel_spam = $data['count_spam'] / $texts_spam;
			
		$rating = $rel_spam / ($rel_ham + $rel_spam);
		
		# Calculate the better probability proposed by Mr. Robinson
		$all = $data['count_ham'] + $data['count_spam'];
		return (($this->config['rob_s'] * $this->config['rob_x']) + ($all * $rating)) / ($this->config['rob_s'] + $all);
		
	}
	
	/**
	 * Check the validity of the category of a request
	 *
	 * @access private
	 * @param string $category
	 * @return void
	 */
	
	private function _check_category($category)
	{
		return $category === self::HAM or $category === self::SPAM;
	}
	
	/**
	 * Learn a reference text
	 *
	 * @access public
	 * @param string $text
	 * @param const $category Either b8::SPAM or b8::HAM
	 * @return mixed void or an error code
	 */
	
	public function learn($text = NULL, $category = NULL)
	{
	
		# Let's first see if the user called the function correctly
		if($text === NULL)
			return self::TRAINER_TEXT_MISSING;
		if($category === NULL)
			return self::TRAINER_CATEGORY_MISSING;
			
		return $this->_process_text($text, $category, self::LEARN);
		
	}
	
	/**
	 * Unlearn a reference text
	 *
	 * @access public
	 * @param string $text
	 * @param const $category Either b8::SPAM or b8::HAM
	 * @return mixed void or an error code
	 */
	
	public function unlearn($text = NULL, $category = NULL)
	{
	
		# Let's first see if the user called the function correctly
		if($text === NULL)
			return self::TRAINER_TEXT_MISSING;
		if($category === NULL)
			return self::TRAINER_CATEGORY_MISSING;
			
		return $this->_process_text($text, $category, self::UNLEARN);
		
	}
	
	/**
	 * Does the actual interaction with the storage backend for learning or unlearning texts
	 *
	 * @access private
	 * @param string $text
	 * @param const $category Either b8::SPAM or b8::HAM
	 * @param const $action Either b8::LEARN or b8::UNLEARN
	 * @return mixed void or an error code
	 */
	
	private function _process_text($text, $category, $action)
	{
	
		# Look if the request is okay
		if($this->_check_category($category) === FALSE)
			return self::TRAINER_CATEGORY_FAIL;
		
		# Get all tokens from $text
		$tokens = $this->lexer->get_tokens($text);
		
		# Check if the lexer failed
		# (if so, $tokens will be a lexer error code, if not, $tokens will be an array)
		if(!is_array($tokens))
			return $tokens;
		
		# Pass the tokens and what to do with it to the storage backend
		return $this->storage->process_text($tokens, $category, $action);
		
	}
	
}

?>